File: ejemplos/general/04-analizador-de-gastos-csv/ejemplo_csv.go
/*
Ejemplo: Leer un CSV real desde disco, filtrar y calcular sumas

Qué enseña este ejemplo:
  - Cómo abrir y leer un archivo CSV con os.Open y encoding/csv
  - Cómo iterar por cada fila del archivo CSV
  - Cómo convertir strings a números con strconv.ParseFloat
  - Cómo agrupar filas y calcular sumas / promedios usando mapas

Para tu reto (analizador de gastos):
  Aplica el mismo patrón para leer tu CSV de gastos
  (fecha, categoría, monto) y calcular totales por categoría.

Glosario:
  os.Open(archivo)      — Abre un archivo en modo lectura.
  defer f.Close()       — Asegura que el archivo se cerrará cuando termine la función.
  csv.NewReader(f)      — Crea un lector especializado que entiende el formato CSV (comas, comillas).
  strconv.ParseFloat(s) — Convierte un texto como "1234.50" a un número flotante (decimal).
*/

package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strconv"
)

type Empleado struct {
	Nombre       string
	Departamento string
	Salario      float64
}

func leerEmpleados(rutaCsv string) []Empleado {
	file, err := os.Open(rutaCsv)
	if err != nil {
		fmt.Printf("No se pudo abrir el archivo: %v\n", err)
		return []Empleado{}
	}
	defer file.Close()

	reader := csv.NewReader(file)
	// Leer todas las filas
	filas, err := reader.ReadAll()
	if err != nil {
		fmt.Printf("Error leyendo el CSV: %v\n", err)
		return []Empleado{}
	}

	var empleados []Empleado

	// Iterar omitiendo la cabecera (índice 0)
	for i := 1; i < len(filas); i++ {
		fila := filas[i]
		if len(fila) < 3 {
			continue
		}

		nombre := fila[0]
		departamento := fila[1]
		salario, _ := strconv.ParseFloat(fila[2], 64)

		empleados = append(empleados, Empleado{
			Nombre:       nombre,
			Departamento: departamento,
			Salario:      salario,
		})
	}

	return empleados
}

type Estadisticas struct {
	Suma     float64
	Cantidad int
}

func reportePorDepartamento(empleados []Empleado) {
	totales := make(map[string]*Estadisticas)

	for _, emp := range empleados {
		dept := emp.Departamento
		if totales[dept] == nil {
			totales[dept] = &Estadisticas{}
		}
		totales[dept].Suma += emp.Salario
		totales[dept].Cantidad++
	}

	fmt.Println("\n--- Reporte por departamento ---")
	for dept, stats := range totales {
		promedio := stats.Suma / float64(stats.Cantidad)
		// %.0f imprime sin decimales
		fmt.Printf("  %s: total=$%.0f | promedio=$%.0f | empleados=%d\n", dept, stats.Suma, promedio, stats.Cantidad)
	}
}

func main() {
	archivo := "empleados.csv"

	// Crear archivo de ejemplo si no existe
	if _, err := os.Stat(archivo); os.IsNotExist(err) {
		datos := []string{
			"nombre,departamento,salario",
			"Ana García,Ingeniería,35000",
			"Luis Martínez,Marketing,28000",
			"Carmen López,Ingeniería,42000",
			"Pedro Sánchez,Marketing,31000",
			"Sofía Ramírez,Ingeniería,39000",
		}
		
		f, _ := os.Create(archivo)
		for _, linea := range datos {
			f.WriteString(linea + "\n")
		}
		f.Close()
		fmt.Printf("Archivo '%s' creado de ejemplo.\n", archivo)
	}

	empleados := leerEmpleados(archivo)
	fmt.Printf("\nSe leyeron %d empleados.\n", len(empleados))
	reportePorDepartamento(empleados)
}

File: archivo-historico/ejemplos/general/04-analizador-de-gastos-csv/ejemplo_csv.rs
/*
Ejemplo: Leer un CSV real desde disco, filtrar y calcular sumas

Qué enseña este ejemplo:
  - Cómo usar la biblioteca estándar para leer archivos estructurados.
  - Cómo convertir strings a números con .parse()
  - Cómo agrupar usando un HashMap

Para tu reto (analizador de gastos):
  Aplica el mismo patrón para leer tu CSV de gastos
  (fecha, categoría, monto) y calcular totales por categoría.

Glosario:
  File::open(ruta)     — Abre un archivo en modo lectura.
  .parse::<f64>()      — Intenta convertir un string a un número decimal (float de 64 bits).
  HashMap::new()       — Crea una estructura de diccionario para guardar totales por clave.
  .entry(clave)        — Busca una clave en el map para modificarla o insertarla si no existe.
*/

use std::collections::HashMap;
use std::fs::File;
use std::io::Write;
use std::path::Path;

struct Empleado {
    _nombre: String,
    departamento: String,
    salario: f64,
}

struct Estadisticas {
    suma: f64,
    cantidad: i32,
}

fn leer_empleados(ruta_csv: &str) -> Vec<Empleado> {
    let mut empleados = Vec::new();

    let file = match File::open(ruta_csv) {
        Ok(f) => f,
        Err(_) => {
            println!("Archivo no encontrado: {}", ruta_csv);
            return empleados;
        }
    };

    use std::io::{BufRead, BufReader};
    let reader = BufReader::new(file);

    for (index, line) in reader.lines().enumerate() {
        if index == 0 {
            continue; // Saltar cabecera
        }
        if let Ok(l) = line {
            let cols: Vec<&str> = l.split(',').collect();
            if cols.len() >= 3 {
                let nombre = cols[0].trim().to_string();
                let departamento = cols[1].trim().to_string();
                let salario = cols[2].trim().parse::<f64>().unwrap_or(0.0);

                empleados.push(Empleado {
                    _nombre: nombre,
                    departamento,
                    salario,
                });
            }
        }
    }

    empleados
}

fn reporte_por_departamento(empleados: &[Empleado]) {
    let mut totales: HashMap<String, Estadisticas> = HashMap::new();

    for emp in empleados {
        let entry = totales
            .entry(emp.departamento.clone())
            .or_insert(Estadisticas { suma: 0.0, cantidad: 0 });
        entry.suma += emp.salario;
        entry.cantidad += 1;
    }

    println!("\n--- Reporte por departamento ---");
    for (dept, stats) in totales.iter() {
        let promedio = stats.suma / stats.cantidad as f64;
        println!(
            "  {}: total=${} | promedio=${} | empleados={}",
            dept, stats.suma, promedio, stats.cantidad
        );
    }
}

fn main() {
    let archivo = "empleados.csv";

    if !Path::new(archivo).exists() {
        let mut file = File::create(archivo).expect("No se pudo crear el archivo");
        let datos = "\
nombre,departamento,salario
Ana García,Ingeniería,35000
Luis Martínez,Marketing,28000
Carmen López,Ingeniería,42000
Pedro Sánchez,Marketing,31000
Sofía Ramírez,Ingeniería,39000
";
        file.write_all(datos.as_bytes()).unwrap();
        println!("Archivo '{}' creado de ejemplo.", archivo);
    }

    let empleados = leer_empleados(archivo);
    println!("\nSe leyeron {} empleados.", empleados.len());
    reporte_por_departamento(&empleados);
}
